### DISCUSSION TITLE
PDK response context cannot replace body_filter

### DISCUSSION THREAD

**Discussion by provMichaelGugino** [Kong/kong#11545] [Help]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Kong version (`$ kong version`)

3.2.2

### Current Behavior

When following documentation here: https://docs.konghq.com/gateway/latest/plugin-development/custom-logic/#available-contexts

It notes that response context is to replace header_filter and body_filter.

When I write a response plugin that attempts to access or set raw body:
```
function plugin:response(plugin_conf)
  local status = kong.response.get_status()
  -- Error on below line
  original_body = kong.response.get_raw_body()
  ...
```

I get the following error message:
```
2023/07/17 16:00:25 [error] 101#0: *154 lua entry thread aborted: runtime error: /usr/local/share/lua/5.1/kong/pdk/private/phases.lua:90: function cannot be called in response phase (only in: body_filter)
...
```

### Expected Behavior

All body filter functions should be available in response context.

### Steps To Reproduce

1. Write kong plugin using lua containing response context method
2. Use method such as kong.response.get_raw_body()
3. Execute plugin

**Comment by mfeledyn** [comment]
I see same behaviour in Kong 3.3. This is a **blocking point**.
I tried many things. I cannot use kong.response.get_raw_body() in response function. Trying to get the body using ngx.arg[1]  generates an error (tells the phase is "access" which is a wrong phase for the call, but it's not called from the access phase ?...)
I checked phase in functions:
- in header_filter function, ngx.get_phase() returns "header_filter" (phase OK)
- in body_filter function, ngx.get_phase() returns "body_filter" (phase OK)
- in response function, ngx.get_phase() returns "access" (phase KO, it's wrong)

So it seems that response function is not taken into account by both Kong layers, and Nginx also.

One more thing, When I use kong.service.response.get_body() I intend to get the response body, but actually I get the request body !?!...

**Comment by StarlightIbuki** [comment]
@mfeledyn thanks for the report. Could you share a reproducible example of the issue? Especially the `kong.service.response.get_body()`'s unusual behavior.

**Comment by mfeledyn** [comment]
> @mfeledyn thanks for the report. Could you share a reproducible example of the issue? Especially the `kong.service.response.get_body()`'s unusual behavior.

Here is the code of a short plugin for showing the problem:

```
local inspect = require "inspect"

local klog = kong.log

local _M = {
  VERSION  = "0.0.1",
  PRIORITY = 1000
}

local log_prefix = "##phasestest##"


function _M:access(config)
  kong.log.info(log_prefix.." 'access' phase")
  kong.log.info(log_prefix.."phase="..tostring(ngx.get_phase()))
  kong.service.request.enable_buffering()

end

function _M:response(config)
  kong.log.info(log_prefix.." 'response' phase")
  kong.log.info(log_prefix.."phase="..tostring(ngx.get_phase()))
  kong.log.info(log_prefix.." - INSPECT(kong.service.response.get_body())="..inspect(kong.service.response.get_body()))

end

function _M:log(config)
  kong.log.info(log_prefix.." 'log' phase")
  kong.log.info(log_prefix.."phase="..tostring(ngx.get_phase()))
end

return _M

```
And the logs:

```
2023/08/31 15:00:45 [info] 1264#0: *388 [kong] handler.lua:14 [phasestest] ##phasestest## 'access' phase, client: 172.22.0.1, server: kong, request: "GET /testresponse/echo HTTP/1.1", host: "localhost:8000"
2023/08/31 15:00:45 [info] 1264#0: *388 [kong] handler.lua:15 [phasestest] ##phasestest##phase=access, client: 172.22.0.1, server: kong, request: "GET /testresponse/echo HTTP/1.1", host: "localhost:8000"
2023/08/31 15:00:45 [info] 1264#0: *388 [kong] handler.lua:21 [phasestest] ##phasestest## 'response' phase while sending to client, client: 172.22.0.1, server: kong, request: "GET /testresponse/echo HTTP/1.1", host: "localhost:8000"
2023/08/31 15:00:45 [info] 1264#0: *388 [kong] handler.lua:22 [phasestest] ##phasestest##phase=access while sending to client, client: 172.22.0.1, server: kong, request: "GET /testresponse/echo HTTP/1.1", host: "localhost:8000"
2023/08/31 15:00:45 [info] 1264#0: *388 [kong] handler.lua:23 [phasestest] ##phasestest## - INSPECT(kong.service.response.get_body())={
  body = "",
  headers = {
    Accept = { "*/*",
      <metatable> = <1>{}
    },
    Connection = { "keep-alive",
      <metatable> = <table 1>
    },
    ["User-Agent"] = { "curl/7.88.1",
      <metatable> = <table 1>
    },
    ["X-Forwarded-For"] = { "172.22.0.1",
      <metatable> = <table 1>
    },
    ["X-Forwarded-Host"] = { "localhost",
      <metatable> = <table 1>
    },
    ["X-Forwarded-Path"] = { "/testresponse/echo",
      <metatable> = <table 1>
    },
    ["X-Forwarded-Port"] = { "8000",
      <metatable> = <table 1>
    },
    ["X-Forwarded-Prefix"] = { "/testresponse",
      <metatable> = <table 1>
    },
    ["X-Forwarded-Proto"] = { "http",
      <metatable> = <table 1>
    },
    ["X-Real-Ip"] = { "172.22.0.1",
      <metatable> = <table 1>
    }
  },
  method = "GET",
  url = "/mickgo/echo"
} while sending to client, client: 172.22.0.1, server: kong, request: "GET /testresponse/echo HTTP/1.1", host: "localhost:8000"
2023/08/31 15:00:45 [info] 1264#0: *388 [kong] handler.lua:28 [phasestest] ##phasestest## 'log' phase while logging request, client: 172.22.0.1, server: kong, request: "GET /testresponse/echo HTTP/1.1", host: "localhost:8000"
2023/08/31 15:00:45 [info] 1264#0: *388 [kong] handler.lua:29 [phasestest] ##phasestest##phase=log while logging request, client: 172.22.0.1, server: kong, request: "GET /testresponse/echo HTTP/1.1", host: "localhost:8000"

```

Obviously the inspect(kong.service.response.get_body() shows a request attributes.
And you can see in the log that in "response" code, the call to "ngx.get_phase()" reports 'access'.

**Comment by hanshuebner** [comment]
As [documented](https://docs.konghq.com/gateway/latest/plugin-development/custom-logic/#available-contexts), the `response` function is called in the `access` phase.  If you need to call `kong.response.get_raw_body()`, you'll need to implement the `body_filter` function and call it from there.

**Comment by mfeledyn** [comment]
> As [documented](https://docs.konghq.com/gateway/latest/plugin-development/custom-logic/#available-contexts), the `response` function is called in the `access` phase. If you need to call `kong.response.get_raw_body()`, you'll need to implement the `body_filter` function and call it from there.

Indeed it is documented (thx for the link) as in the `access `phase. Strange since it is written: 
_"Replaces both `header_filter()` and `body_filter()`. Executed after the whole response has been received from the upstream service, but before sending any part of it to the client"_.
If the response was received, it seems to me that it's not the access phase anymore.
I think I have read something telling that `response` phase enables changing both headers and body, something that cannot be done in `header_filter `or `body_filter`. Sometimes you have to analyse the body of the response and then change/replace it, but you have to modify Content-length header and Content-type, the later cannot be done in`body_filter()` .

Still the `kong.**service**.response.get_body()` (in the code, it is not `kong.response.get_raw_body()`) gives the service request, not the service response. I get the same in  `header_filter `or `body_filter`.

Btw if I try to implement `response()` while keeping `header_filter()` and/or `body_filter()` I get:
`nginx: [error] init_by_lua error: /usr/local/share/lua/5.1/kong/init.lua:621: error loading plugin schemas: on plugin 'phasestest': Plugin "phasestest" can't be loaded because it implements both `response` and `header_filter` or `body_filter` methods.`
Seems good, but if it's not the same phase I don't understand ?...

**Comment by hanshuebner** [comment]
When talking about phases in the documentation, we're almost always talking about the phases defined by OpenResty, which in turn (mostly) map to [nginx's http phases](http://nginx.org/en/docs/dev/development_guide.html#http_phases).  Plugin callback functions are always called in one of the available OpenResty.phases and some of the callbacks actually share their name with a phase defined by nginx or, but this is not true for all of them.

As you have observed, `response()` cannot be used together with `header_filter()` and/or `body_filter()`.  Is that something that is blocking what you want to do?

Regarding `kong.service.response.get_body()`, is the echo service that you're using maybe echoing the complete request, including the headers?

**Comment by mfeledyn** [comment]
Hello, thank you for answering.
You are right, the echo service I use is sending back my request as a JSON object in the response. I feel stupid. 

I can get the body using `get_raw_body()`. The `get_body()` returns nil because it does not find how to decode the response body.
The fact that I cannot use `response()` together with `header_filter()` and `body_filter()` is not blocking, I understand the logic on this.
What I'd like to do is be able to modified both headers and body at the same time, so before sending anything to the client.

Nevertheless another thing surprises me: I cannot list headers using object return by kong.service.response.get_headers().

`inspect(kong.service.response.get_headers())` gives: 
```
{
  <metatable> = {
    __index = <function 1>
  }
}
```

`(#kong.service.response.get_headers())` gives `0`

```
    for k, v in pairs(kong.service.response.get_headers()) do
      kong.log.info(log_prefix().."header: key="..k..", value="..json.encode(v))
    end
    for k, v in ipairs(kong.service.response.get_headers()) do
      kong.log.info(log_prefix().."header: key="..k..", value="..json.encode(v))
    end
```
The code above does not log anything.

There are headers in server response, for example this call `inspect(kong.service.response.get_headers()["Server"])` gives `"My GO test server"` (I'm using this test server indeed...)

Would you know a way to iterate over service response headers ?

**Comment by hanshuebner** [comment]
> Would you know a way to iterate over service response headers ?

From looking at the code, it seems that there currently is no way, and I think this is a bug - After all, the function is [documented](https://docs.konghq.com/gateway/latest/plugin-development/pdk/kong.service.response/) to return a Lua table with the headers, but it does not seem as if it really lives up to that promise.  Would you have time to open an issue with a repro case?

**Comment by mfeledyn** [comment]
> > Would you know a way to iterate over service response headers ?
> 
> From looking at the code, it seems that there currently is no way, and I think this is a bug - After all, the function is [documented](https://docs.konghq.com/gateway/latest/plugin-development/pdk/kong.service.response/) to return a Lua table with the headers, but it does not seem as if it really lives up to that promise. Would you have time to open an issue with a repro case?

Ok I'll try to open an issue on this.
Btw I have tested and I can get the body (using `get_raw_body()`) in `header_filter()` function.
I'll try using the response() function to tinker with both headers and body at the same time (this was my initial goal :-) ).

**Comment by mfeledyn** [comment]
Issue made from this case: https://github.com/Kong/kong/issues/11546

**Comment by mfeledyn** [comment]
> I'll try using the response() function to tinker with both headers and body at the same time (this was my initial goal :-) ).

This session was successful ^^. In `header_filter()` I can use `kong.service.response.get_raw_body()` to access the body and then analyse it. I can make a modified body (to be used later in `body_filter()`) and set "Content-length" header to the modified body length. So I think I'm good now regarding my initial goal. 
Nevertheless it would be useful to be able to iterate over headers ;-)

Thanks so much for your answers, they helped me a lot.
